sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Expression, ExprId, NamedExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, Partial, PartialMerge}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Reassigns expression IDs in every expression of the given LogicalPlan (including subqueries
+ * contained in the plan).
+ */
+object AssignNewExprIds extends Rule[LogicalPlan] {
+
+  private def assignNewExprIdsinExpr(input: Expression,
+                                     reassignedExprs: mutable.HashMap[ExprId, Attribute]):
+  Expression = input match {
+    case a: Attribute =>
+      val newAttribute = reassignedExprs.getOrElse(a.exprId,
+        a.withExprId(NamedExpression.newExprId))
+      reassignedExprs.put(a.exprId, newAttribute)
```

Comment:
can we avoid the `put` if the key was actually present in the map? e.g.

```
reassignedExprs.get(a.exprId).getOrElse {
  val newAttribute = a.withExprId(NamedExpression.newExprId)
  reassignedExprs.put(a.exprId, newAttribute)
  reassignedExprs.put(newAttribute.exprId, newAttribute)
  newAttribute
}
```

same for the `Alias` case below.

Reply:
neat!
I think with Alias, we always have to `put` as we always update the expression id.